Specifying a custom config.resource causes a "resource not found on classpath" in Play 2.8

### Play Version

2.8.0

### API

Scala

### Operating System

Linux bob 5.3.0-24-generic #26-Ubuntu SMP Thu Nov 14 01:33:18 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

### JDK

java version "1.8.0_221"
Java(TM) SE Runtime Environment (build 1.8.0_221-b11)
Java HotSpot(TM) 64-Bit GraalVM EE 19.1.1 (build 25.221-b11-jvmci-19.1-b01, mixed mode)

### Library Dependencies

N/A

### Expected Behavior

1. I have a file `test.conf` in my `conf/` folder, next to application.conf
2. I specify `-Dconfig.resource=test.conf` when running SBT
3. The application uses `test.conf`

### Actual Behavior

1. I have a file `test.conf` in my `conf/` folder, next to application.conf
2. I specify `-Dconfig.resource=test.conf` when running SBT
3. **The application errors saying test.conf cannot be found on the classpath**

I strongly suspect this to be the cause but would like a second opinion, maybe from @dwijnand ?: https://github.com/playframework/playframework/commit/1b69712bb4df23f7dd6b194e06dbc3c96c2617ed#diff-9c80b5c4ff3925eebc1aa4fab735f97fL73

### Reproducible Test Case

Clone: https://github.com/henricook/play-2.8-bug-config-resource
**Additional information is in the README**

See the regression:
-  `cd play-scala-seed`
- `ls conf` (observe 'test.conf' exists)
- `sbt -Dsbt.version=1.3.5 -Dconfig.resource=test.conf clean run` (observe error)

-----

See it working on 2.7.x:
- `cd ../play-scala-seed-2.7.x`
- `ls conf` (observe 'test.conf' exists)
- `sbt -Dsbt.version=1.3.5 -Dconfig.resource=test.conf clean run` (observe successful start that proves regression)

`sbt.version` included only to ensure a level playing field for this test

### Error dump

```
[error] Configuration error: Configuration error[test.conf: java.io.IOException: resource not found on classpath: test.conf]
[error]         at play.api.Configuration$.configError(Configuration.scala:138)
[error]         at play.api.Configuration$.load(Configuration.scala:81)
[error]         at play.core.server.DevServerStart$.$anonfun$mainDev$1(DevServerStart.scala:230)
[error]         at play.utils.Threads$.withContextClassLoader(Threads.scala:21)
[error]         at play.core.server.DevServerStart$.mainDev(DevServerStart.scala:75)
[error]         at play.core.server.DevServerStart$.mainDevHttpMode(DevServerStart.scala:49)
[error]         at play.core.server.DevServerStart.mainDevHttpMode(DevServerStart.scala)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498)
[error]         at play.runsupport.Reloader$.startDevMode(Reloader.scala:306)
[error]         at play.sbt.run.PlayRun$.devModeServer$lzycompute$1(PlayRun.scala:98)
[error]         at play.sbt.run.PlayRun$.devModeServer$1(PlayRun.scala:81)
[error]         at play.sbt.run.PlayRun$.$anonfun$playRunTask$3(PlayRun.scala:105)
[error]         at play.sbt.run.PlayRun$.$anonfun$playRunTask$3$adapted(PlayRun.scala:67)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] Caused by: com.typesafe.config.ConfigException$IO: test.conf: java.io.IOException: resource not found on classpath: test.conf
[error]         at com.typesafe.config.impl.Parseable.parseValue(Parseable.java:190)
[error]         at com.typesafe.config.impl.Parseable.parseValue(Parseable.java:174)
[error]         at com.typesafe.config.impl.Parseable.parse(Parseable.java:301)
[error]         at com.typesafe.config.ConfigFactory.parseResources(ConfigFactory.java:1058)
[error]         at com.typesafe.config.ConfigFactory.parseResources(ConfigFactory.java:986)
[error]         at com.typesafe.config.DefaultConfigLoadingStrategy.parseApplicationConfig(DefaultConfigLoadingStrategy.java:49)
[error]         at com.typesafe.config.ConfigFactory.defaultApplication(ConfigFactory.java:529)
[error]         at play.api.Configuration$.load(Configuration.scala:58)
[error]         at play.core.server.DevServerStart$.$anonfun$mainDev$1(DevServerStart.scala:230)
[error]         at play.utils.Threads$.withContextClassLoader(Threads.scala:21)
[error]         at play.core.server.DevServerStart$.mainDev(DevServerStart.scala:75)
[error]         at play.core.server.DevServerStart$.mainDevHttpMode(DevServerStart.scala:49)
[error]         at play.core.server.DevServerStart.mainDevHttpMode(DevServerStart.scala)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498)
[error]         at play.runsupport.Reloader$.startDevMode(Reloader.scala:306)
[error]         at play.sbt.run.PlayRun$.devModeServer$lzycompute$1(PlayRun.scala:98)
[error]         at play.sbt.run.PlayRun$.devModeServer$1(PlayRun.scala:81)
[error]         at play.sbt.run.PlayRun$.$anonfun$playRunTask$3(PlayRun.scala:105)
[error]         at play.sbt.run.PlayRun$.$anonfun$playRunTask$3$adapted(PlayRun.scala:67)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] Caused by: java.io.IOException: resource not found on classpath: test.conf
[error]         at com.typesafe.config.impl.Parseable$ParseableResources.rawParseValue(Parseable.java:726)
[error]         at com.typesafe.config.impl.Parseable$ParseableResources.rawParseValue(Parseable.java:701)
[error]         at com.typesafe.config.impl.Parseable.parseValue(Parseable.java:180)
[error]         at com.typesafe.config.impl.Parseable.parseValue(Parseable.java:174)
[error]         at com.typesafe.config.impl.Parseable.parse(Parseable.java:301)
[error]         at com.typesafe.config.ConfigFactory.parseResources(ConfigFactory.java:1058)
[error]         at com.typesafe.config.ConfigFactory.parseResources(ConfigFactory.java:986)
[error]         at com.typesafe.config.DefaultConfigLoadingStrategy.parseApplicationConfig(DefaultConfigLoadingStrategy.java:49)
[error]         at com.typesafe.config.ConfigFactory.defaultApplication(ConfigFactory.java:529)
[error]         at play.api.Configuration$.load(Configuration.scala:58)
[error]         at play.core.server.DevServerStart$.$anonfun$mainDev$1(DevServerStart.scala:230)
[error]         at play.utils.Threads$.withContextClassLoader(Threads.scala:21)
[error]         at play.core.server.DevServerStart$.mainDev(DevServerStart.scala:75)
[error]         at play.core.server.DevServerStart$.mainDevHttpMode(DevServerStart.scala:49)
[error]         at play.core.server.DevServerStart.mainDevHttpMode(DevServerStart.scala)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498)
[error]         at play.runsupport.Reloader$.startDevMode(Reloader.scala:306)
[error]         at play.sbt.run.PlayRun$.devModeServer$lzycompute$1(PlayRun.scala:98)
[error]         at play.sbt.run.PlayRun$.devModeServer$1(PlayRun.scala:81)
[error]         at play.sbt.run.PlayRun$.$anonfun$playRunTask$3(PlayRun.scala:105)
[error]         at play.sbt.run.PlayRun$.$anonfun$playRunTask$3$adapted(PlayRun.scala:67)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] stack trace is suppressed; run last Compile / run for the full output
[error] (Compile / run) java.lang.reflect.InvocationTargetException
```

--- a/core/play/src/main/scala/play/api/Configuration.scala
+++ b/core/play/src/main/scala/play/api/Configuration.scala
@@ -4,6 +4,7 @@
 
 package play.api
 
+import java.io.File
 import java.net.URI
 import java.net.URL
 import java.util.Properties
@@ -52,10 +53,23 @@ object Configuration {
 
       // Resolve application.conf
       val applicationConfig: Config = {
-        val parseOptions = ConfigParseOptions.defaults
-          .setClassLoader(classLoader)
-          .setAllowMissing(allowMissingApplicationConf)
-        ConfigFactory.defaultApplication(parseOptions)
+        // The additional config.resource/config.file logic exists because
+        // ConfigFactory.defaultApplication will blow up if those are defined but the file is missing
+        // despite "setAllowMissing" (see DefaultConfigLoadingStrategy).
+        // In DevMode this is relevant for config.resource as reloader.currentApplicationClassLoader
+        // is null at the start of 'run', so the application classpath isn't available, which means
+        // the resource will be missing.  For consistency (and historic behaviour) do config.file too.
+        {
+          sys.props.get("config.resource").map(resource => ConfigFactory.parseResources(classLoader, resource))
+        }.orElse {
+            sys.props.get("config.file").map(fileName => ConfigFactory.parseFileAnySyntax(new File(fileName)))
+          }
+          .getOrElse {
+            val parseOptions = ConfigParseOptions.defaults
+              .setClassLoader(classLoader)
+              .setAllowMissing(allowMissingApplicationConf)
+            ConfigFactory.defaultApplication(parseOptions)
+          }
       }
 
       // Resolve another .conf file so that we can override values in Akka's

--- a/core/play/src/test/scala/play/api/ConfigurationSpec.scala
+++ b/core/play/src/test/scala/play/api/ConfigurationSpec.scala
@@ -357,7 +357,7 @@ class ConfigurationSpec extends Specification {
 
       val direct = Map(
         "config.resource" -> "application.from-direct.res.conf",
-        "config.resource" -> "application.from-direct.file.conf",
+        "config.file"     -> "application.from-direct.file.conf",
       )
 
       val cl = new InMemoryResourceClassLoader(

